src/bookshelf/src/bksw_sec.erl

```diff
@@ -35,109 +37,283 @@ is_authorized(Req0, #context{} = Context) ->
+header_auth(RequestId, IncomingAuth, Req0, Context, Headers0) ->
+    try
+        (ParseAuth = parse_authorization(IncomingAuth)) /= err orelse throw({RequestId, Req0, Context}),
+        [Credential, SignedHeaderKeysString, IncomingSignature] = ParseAuth,
+        XAmzDate = wrq:get_req_header("x-amz-date", Req0),
+        common_auth(RequestId, Req0, Context, Credential, XAmzDate, SignedHeaderKeysString, IncomingSignature, "300", Headers0, authorization_header)
+    catch
+        _ -> encode_access_denied_error_response(RequestId, Req0, Context)
     end.
 
-make_signed_url_authorization(SecretKey, Method, Path, Expires, Headers) ->
+common_auth(RequestId, Req0, #context{reqid = ReqId} = Context, Credential, XAmzDate, SignedHeaderKeysString, IncomingSignature, XAmzExpiresString, Headers0, VerificationType) ->
     try
-        mini_s3:make_signed_url_authorization(SecretKey,
-                                              erlang:list_to_existing_atom(Method),
-                                              Path,
-                                              Expires,
-                                              Headers)
+        (Host = wrq:get_req_header("Host", Req0)) /= undefined orelse throw({RequestId, Req0, Context}),
+
+        (ParseCred = parse_x_amz_credential(Credential)) /= err orelse throw({RequestId, Req0, Context}),
+        [AWSAccessKeyId, CredentialScopeDate, Region | _] = ParseCred,
+
+        % https://docs.aws.amazon.com/general/latest/gr/sigv4-date-handling.html
+        DateIfUndefined = wrq:get_req_header("date", Req0),
+        (Date = get_check_date(XAmzDate, DateIfUndefined, CredentialScopeDate)) /= err orelse throw({RequestId, Req0, Context}),
+
+        AccessKey = bksw_conf:access_key_id(Context),
+        SecretKey = bksw_conf:secret_access_key(Context),
+
+        Headers = process_headers(Headers0),
+
+        SignedHeaderKeys = parse_x_amz_signed_headers(SignedHeaderKeysString),
+        SignedHeaders = get_signed_headers(SignedHeaderKeys, Headers, []),
+
+        RawMethod = wrq:method(Req0),
+        Method = list_to_atom(string:to_lower(erlang:atom_to_list(RawMethod))),
+
+        Path  = wrq:path(Req0),
+
+        {BucketName, Key} = get_bucket_key(Path),
+
+        Config = mini_s3:new(AccessKey, SecretKey, Host),
+        AltHost = mini_s3:get_host_toggleport(Host, Config),
+        AltSignedHeaders = [case {K, V} of {"host", _} -> {"host", AltHost}; _ -> {K, V} end || {K, V} <- SignedHeaders],
+        XAmzExpires = list_to_integer(XAmzExpiresString),
+
+        Sig1 = case VerificationType of
+            presigned_url ->
+
+                "AWS4-HMAC-SHA256" == wrq:get_qs_value("X-Amz-Algorithm", Req0) orelse throw({RequestId, Req0, Context}),
+
+                % temporarily disabling this - should be re-enabled later
+                %true == check_signed_headers_common(SignedHeaders, Headers) orelse throw({RequestId, Req0, Context}),
```

Comment:
Makes me a bit nervous that everything is passing with this commented out.  Might imply we are missing a test or two.

Reply:
The code has now been uncommented (activated).  Things seem to be working fine.